### PR TITLE
feat(frontend): add `StreamProjectMaterialize` PlanNode

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
Add `StreamProjectMaterialize` PlanNode

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [X] My PR **DOES NOT** contain user-facing changes.

## References
https://github.com/risingwavelabs/rfcs/pull/24

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_project.rs [hunk 1]
```diff
@@ -270,23 +270,8 @@ impl ToStream for LogicalProject {
         let new_input = self
             .input()
             .to_stream_with_dist_required(&input_required, ctx)?;
-        let new_logical = self.clone_with_input(new_input.clone());
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_project.rs:273]: Note for review: we already do project merge rule

**Comment by jon-chuang** [comment]
> as accepted first to make sure consensus is reached.

Agreed, I decided to get the ball rolling as I think the consensus is more or less in this direction and I'd like this feature for my demo :)

**Comment by st1page** [comment]
> > as accepted first to make sure consensus is reached.
> 
> Agreed on waiting, I decided to get the ball rolling as I think the consensus is more or less in this direction - hoping to get it sooner as I'd like this feature for my demo :)

I agree that this PR can achieve that with minimal invasiveness and easy to revert it... So I think I will be glad to approve it if it is really urgent. But at least, I think we need to add warnings to users "it is an experimental feature" when we write a UDF on an updatable stream.

**Comment by jon-chuang** [comment]
> But at least, I think we need to add warnings to users "it is an experimental feature" when we write a UDF on an updatable stream.

Agreed.

**Pull request closed without merging**